### PR TITLE
Add products route

### PR DESCRIPTION
**Pull request by van-fs**
This PR adds a products route:
- Products are backed by a static json file in `public/data`
- There are two routes added `products` and `products/<id>`
- [Lunr](https://lunrjs.com/) provides a lightweight search service

On a design note, I moved the getProducts and getProduct middleware out of ProductController.  As I was working, it felt strange to have instance methods inside the ProductController making static calls.  I'll submit another PR to refactor the PingController and cleanup if there are no issues with this design change.

### CODE REVIEW

File: src/controllers/ProductController.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
+import { Request, Response } from 'express';
+import * as fs from 'fs';
+import * as lunr from 'lunr';
+import { Product } from '../models';
+
+export class ProductController {
```
Comment (patrick-fs) [inline src/controllers/ProductController.ts]: This is better named as `ProductDatabase` than `ProductController`

**Comment by patrick-fs** [comment]
> On a design note, I moved the getProducts and getProduct middleware out of ProductController. As I was working, it felt strange to have instance methods inside the ProductController making static calls.

Please see my suggestion - I think you can refactor stuff out of controller and into a `ProductDatabase` class that is a client to the `ProductController`.

File: src/controllers/ProductController.ts (hunk 1 above)
Comment (van-fs) [reply src/controllers/ProductController.ts]: Agreed; created a `ProductCatalog` service and moved it out of the ProductController.

File: src/controllers/ProductController.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * Gets a product based on its ID, which is provided as a route param.
+ * @param req Express Request
+ * @param res Express Response
+ */
+export function getProduct(req: Request, res: Response) {
+  const { id } = req.params;
+
+  console.log(`Getting product with id ${id}`);
+
+  const controller = ProductCatalog.getInstance();
```
Comment (patrick-fs) [inline src/controllers/ProductController.ts]: Nit: you probably shouldn't call the instance `controller` here

File: src/controllers/ProductController.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ * If no search is specified, all products are returned.
+ * @param req Express Request
+ * @param res Express Response
+ */
+export function getProducts(req: Request, res: Response) {
+  const { q } = req.query;
+
+  console.log(`Getting product with query ${q}`);
+
+  const controller = ProductCatalog.getInstance();
```
Comment (patrick-fs) [inline src/controllers/ProductController.ts]: same nit as above.

File: src/controllers/ProductController.ts (hunk 2 above)
Comment (van-fs) [reply src/controllers/ProductController.ts]: ah! yes!

**Pull request merged**